### PR TITLE
WIP - NuPIC.cpp ARM64 build on CircleCI

### PR DESCRIPTION
**Pull request by brev**
Work in Progress.
WIP - NuPIC ARM64 build on CircleCI
https://github.com/htm-community/nupic.cpp/issues/134

* Moving CircleCI from old 1.x config style (`./circle.yml`) to the
  newer 2.x style (`./.circleci/config.yml`).
* Adding a 2nd job to the existing CircleCI config (Darwin), to try and
  build an ARM64 Docker build.
* Updates to Dockerfile for `nupic.cpp` instead of older `nupic.core`.
* Removed empty orphan space chars in config file

With these changes (so far), one can start a local CircleCI build of
NuPIC Docker on ARM64:

```bash
# developing on mac os/x
brew install circleci
cd nupic.cpp/
circleci local execute --job arm64
```

We get to our first error (which seems like the kind of error I'd expect going
from `x86_64` -> `arm64`) below. I have no idea what to do with it, though:

```
[ 63%] Building CXX object src/test/CMakeFiles/unit_tests.dir/unit/ntypes/BasicTypeTest.cpp.o
/usr/local/src/nupic.cpp/src/test/unit/ntypes/BasicTypeTest.cpp:115:49: error: narrowing conversion of '-128' from 'int' to 'nupic::Byte {aka char}' inside { } [-Wnarrowing]
   Byte bufByte[8] = {0, 1, 2, 3, 4, 5, -128, 127};
                                                 ^
src/test/CMakeFiles/unit_tests.dir/build.make:521: recipe for target 'src/test/CMakeFiles/unit_tests.dir/unit/ntypes/BasicTypeTest.cpp.o' failed
make[2]: *** [src/test/CMakeFiles/unit_tests.dir/unit/ntypes/BasicTypeTest.cpp.o] Error 1
CMakeFiles/Makefile2:436: recipe for target 'src/test/CMakeFiles/unit_tests.dir/all' failed
make[1]: *** [src/test/CMakeFiles/unit_tests.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
setup.py: Calling /usr/local/src/nupic.cpp/bindings/py/packaging/setup.py
```

cc @breznak

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -115,8 +106,29 @@ jobs:
             #tar -zcv -f nupic_core-${CIRCLE_SHA1}-darwin64.tar.gz dist
             #aws s3 cp nupic_core-${CIRCLE_SHA1}-darwin64.tar.gz s3://artifacts.numenta.org/numenta/nupic.core/circle/
 
+  # job: ARM64 (arm64)
+  arm64:
+    docker:
+      - image: circleci/buildpack-deps:18.10
```
Comment (breznak) [inline .circleci/config.yml]: can you please add a comment/link where these images come from? Ie how we update in the future

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,5 +1,7 @@
 version: 2
 jobs:
+
+  # job: Darwin (build-and-test)
```
Comment (breznak) [inline .circleci/config.yml]: nit, please add OSX/Darwin here, so people are not confused

File: .circleci/config.yml [hunk 3]
```diff
@@ -115,8 +106,29 @@ jobs:
...
 
+  # job: ARM64 (arm64)
+  arm64:
+    docker:
+      - image: circleci/buildpack-deps:18.10
+    steps:
+      - run:
+          command: |
+            if [[ $CIRCLE_SHELL_ENV == *"localbuild"* ]]; then
+              echo "This is a local build. Enabling sudo for docker"
```
Comment (breznak) [inline .circleci/config.yml]: i guess this will go away for the CI (not local)?

**Comment by brev** [comment]
Closing as this thread is now deprecated, please go here for updates:
https://github.com/htm-community/nupic.cpp/pull/430

**Pull request closed without merging**